src/commcare_cloud/ansible/plugins/lookup/cchq_aws_secret.py

```diff
@@ -0,0 +1,91 @@
+import json
+import logging
+import os
+import cPickle
+import time
+
+from ansible.plugins.lookup import aws_secret
+from cryptography.fernet import Fernet, InvalidToken
+
+__metaclass__ = type
+
+
+DOCUMENTATION = aws_secret.DOCUMENTATION
+RETURN = aws_secret.RETURN
+
+LOOKUP_IN_PROGRESS_ON_ANOTHER_FORK = b'FETCHING'
+
+
+class LookupModule(aws_secret.LookupModule):
+    """
+    commcare-cloud's wrapper around aws_secret
+
+    Adds commcare-cloud specific caching to aws_secret lookup plugin for big speedup
+    Also assumes the value is json and returns the string json-decoded unlike aws_secret.
+
+    To cache even between forks of the same process/run, we use a one-time process-wide key
+    to encrypt values and write them to disk in files <env_dir>/.generated/.
+    After the run is over, the encryption key is lost to time and the files are useless
+    and indecipherable even by the person who ran it.
+    """
+    def run(self, terms, variables, **kwargs):
+        terms = tuple(terms)
+        term, = terms
+        value = self.get_cache(term, inventory_dir=variables['inventory_dir'])
+        try:
+            if value is Ellipsis:
+                logging.debug('Fetching secret {}'.format(terms))
+                try:
+                    value = super(LookupModule, self).run(terms, variables, **kwargs)
+                except Exception as e:
+                    logging.debug('Caching error fetching secret: {}'.format(e))
+                    value = e
+                self.set_cache(term, value, inventory_dir=variables['inventory_dir'])
+        except Exception as e:
+            logging.warn('There was an error fetching or caching the secret: {}'.format(e))
+            raise
+
+        if isinstance(value, Exception):
+            raise value
+        else:
+            return [json.loads(item) for item in value]
```

Comment:
Where is the corresponding `json.dumps` for this `json.loads`?

Reply:
I see. Would it be possible to use `aws_secret.LookupModule` here rather than `AwsSecretsBackend` (which does the `loads` in `_get_secret`)? That would make the symmetry more obvious.

Reply:
We offlined. The issue is that here https://github.com/dimagi/commcare-cloud/pull/4155/files#diff-2898f3b64f32105af2b0b4f8e1c92617R33 where the lookup expression is inserted to a variable that gets passed into ansible, our ability to conditionally decode json and handle the error that would come up is pretty limited, and I found moving it into here was the only way to get the desired behavior on good and bad inputs. It's possible that if we simplified the different default values secrets variables can take (like by making them all default to `None` rather than some to `None` and some to `''`) then we could move this back out and make this lookup plugin compatible with `aws_secret`.

In the meantime I will just better document the incompatibility through comments.